test/framework/management_cluster.go

```diff
@@ -0,0 +1,120 @@
+// management cluster for e2e testing.
+type InitManagementClusterInput struct {
+	Config
+
+	// Scheme is used to initialize the scheme for the management cluster
+	// client.
+	// Defaults to a new runtime.Scheme.
+	Scheme *runtime.Scheme
+
+	// ComponentGenerators is a list objects that supply additional component
+	// YAML to apply to the management cluster.
+	// Please note this is meant to be used at runtime to add YAML to the
+	// management cluster outside of what is provided by the Components field.
+	// For example, a caller could use this field to apply a Secret required by
+	// some component from the Components field.
+	ComponentGenerators []ComponentGenerator
+
+	// NewManagementClusterFn may be used to provide a custom function for
+	// returning a new management cluster. Otherwise kind.NewCluster is used.
+	NewManagementClusterFn func() (ManagementCluster, error)
+}
+
+// Defaults assigns default values to the object.
+func (c *InitManagementClusterInput) Defaults(ctx context.Context) {
+	c.Config.Defaults()
+	if c.Scheme == nil {
+		c.Scheme = runtime.NewScheme()
+	}
+	if c.NewManagementClusterFn == nil {
+		c.NewManagementClusterFn = func() (ManagementCluster, error) {
+			return kind.NewCluster(ctx, c.ManagementClusterName, c.Scheme)
+		}
+	}
+}
+
+// InitManagementCluster returns a new cluster initialized as a CAPI management
+// cluster.
+func InitManagementCluster(ctx context.Context, input *InitManagementClusterInput) ManagementCluster {
+	By("initializing the management cluster")
+	Expect(input).ToNot(BeNil())
+
+	By("initialzing the management cluster configuration defaults")
+	input.Defaults(ctx)
+
+	By("validating the management cluster configuration")
+	Expect(input.Validate()).To(Succeed())
+
+	By("loading the kubernetes and capi core schemes")
+	TryAddDefaultSchemes(input.Scheme)
+
+	By("creating the management cluster")
+	managementCluster, err := input.NewManagementClusterFn()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(managementCluster).ToNot(BeNil())
+
+	// Load the images.
+	if imageLoader, ok := managementCluster.(ImageLoader); ok {
+		By("management cluster supports loading images")
+		for _, image := range input.Images {
+			By(fmt.Sprintf("loading image %s into the management cluster", image))
+			Expect(imageLoader.LoadImage(ctx, image)).To(Succeed())
+		}
+	}
+
+	// Install the YAML from the component generators.
+	for _, componentGenerator := range input.ComponentGenerators {
+		InstallComponents(ctx, managementCluster, componentGenerator)
+	}
+
+	// Install all components.
+	for _, component := range input.Components {
+		for _, source := range component.Sources {
+			name := component.Name
+			if source.Name != "" {
+				name = fmt.Sprintf("%s/%s", component.Name, source.Name)
+			}
+			source.Name = name
+			InstallComponents(ctx, managementCluster, ComponentGeneratorForComponentSource(source))
+		}
+		for _, waiter := range component.Waiters {
+			switch waiter.Type {
```

Comment:
should there be a default case that tells the user a waiter type is unknown and thus being ignored? (but definitely not failing)

Reply:
I thought about that, but the `Validate` call already rejects an unknown `WaiterType`. We can still be careful and add another check here though.

Reply:
ah perfect, missed that. so long as i'm not wondering why things break when i type `servcie` 😄  i'm fine!